### PR TITLE
[bitnami/logstash] fix: support NodePorts in bitnami logstash chart

### PR DESCRIPTION
**Pull request by cbrwflo**
### Description of the change

Omits the `nodePort` field on the logstash headless service.

This fixes the error `spec.ports[1].nodePort: Forbidden: may not be used when type is 'ClusterIP'` .

### Applicable issues

Fixes:

- #17482
- #15393
- #6278

### Additional information

I didn't see a consistent way of expanding the `service.spec.ports` for headless services across bitnami charts, so I modelled this after the [fluentd headless service](https://github.com/bitnami/charts/blob/main/bitnami/fluentd/templates/aggregator-svc-headless.yaml#L36). 

Instead of using `service.spec.ports.name` (since it's an optional field when only 1 port is used), I opted to use `service.spec.port.port` instead (a required field). I don't love the nested `omit`, and happy to change the PR to "expand" the fields instead if you prefer.

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/logstash/templates/headless-svc.yaml [hunk 1]
```diff
@@ -25,5 +25,8 @@ spec:
   type: ClusterIP
   clusterIP: None
   ports:
-    {{- include "common.tplvalues.render" (dict "value" .Values.service.ports "context" $) | nindent 4 }}
+  {{- range $service := .Values.service.ports }}
+    - port: {{ $service.port }}
+      {{- (omit (omit $service "nodePort") "port") | toYaml | nindent 6 }}
+  {{- end }}
```
Comment (FraPazGal) [inline bitnami/logstash/templates/headless-svc.yaml:31]:
```suggestion
  {{- range $service := .Values.service.ports }}
    - name: {{ $service.name }}
      port: {{ $service.port }}
      targetPort: {{ $service.targetPort }}
      protocol: {{ $service.protocol }}
  {{- end }}
```

A bit more verbose but IMO it's also more intuitive and avoids using the nested `omit`.

**Review by FraPazGal — COMMENTED** [review]
Hello @cbrwflo, thanks for catching this error and creating a PR!

Checking the existing logic, you couldn't find any similar approach because we don't normally use an array for `service.ports`. To fix this, we could also refactor the whole ports logic and use the `service.extraPorts` parameter (see `bitnami/schema-registry` for an example). The problem with that solution is that it would require a new major version for the chart and I don't think it is worthwhile.

For the reasons explained above, I think we can continue with you approach. I've added a suggestion to avoid the `omit` usage, please check it and tell me what you think.

File: bitnami/logstash/templates/headless-svc.yaml [hunk 2]
```diff
@@ -25,5 +25,10 @@ spec:
   type: ClusterIP
   clusterIP: None
   ports:
-    {{- include "common.tplvalues.render" (dict "value" .Values.service.ports "context" $) | nindent 4 }}
+  {{- range $service := .Values.service.ports }}
+    - name: {{ $service.name }}
+      port: {{ $service.port }}
+      targetPort: {{ $service.targetPort }}
+      protocol: {{ $service.protocol }}
+  {{- end }}
```
Comment (juan131) [inline bitnami/logstash/templates/headless-svc.yaml:33]: There's a problem with this implementation: it's not compatible with using templates at `.Values.service.ports` while the previous solution was.

Comment (FraPazGal) [reply bitnami/logstash/templates/headless-svc.yaml:33]: True, we'll be losing that option in this case... I guess we can con with the `omit` solution and work on reworking this logic in a future refactor for this chart.

Comment (juan131) [inline bitnami/logstash/templates/headless-svc.yaml:33]: This should be compatible with using templates.
```suggestion
    {{- range $port := .Values.service.ports }}
    {{- omit (include "common.tplvalues.render" ( dict "value" $port "context" $ ) | fromYaml) "nodePort" | toYaml | nindent 4 }}
    {{- end }}
```

**Comment by carrodher** [comment]
We just reopened this PR, @cbrwflo could you please take a look at the review comments from my teammates? Apart from that, it seems there was a new version during this time, could you please fix the conflicts? Thanks

**Comment by cbrwflo** [comment]
Make fresh! Still interested in this (need to rebase), back from vacay 10/25

**Comment by cbrwflo** [comment]
I intend to work on this

File: bitnami/logstash/templates/headless-svc.yaml [hunk 3]
```diff
@@ -17,6 +17,7 @@ spec:
   type: ClusterIP
   clusterIP: None
   ports:
-    {{- include "common.tplvalues.render" (dict "value" .Values.service.ports "context" $) | nindent 4 }}
-  {{- $podLabels := include "common.tplvalues.merge" ( dict "values" ( list .Values.podLabels .Values.commonLabels ) "context" . ) }}
-  selector: {{- include "common.labels.matchLabels" ( dict "customLabels" $podLabels "context" $ ) | nindent 4 }}
+    {{- range $port := .Values.service.ports }}
+    {{- omit (include "common.tplvalues.render" ( dict "value" $port "context" $ ) | fromYaml) "nodePort" | toYaml | nindent 4 }}
+    {{- end }}
+  selector: {{ include "common.labels.matchLabels" . | nindent 4 }}
```
Comment (juan131) [inline bitnami/logstash/templates/headless-svc.yaml:23]: Please maintain the original way to setup the selector:
```suggestion
  {{- $podLabels := include "common.tplvalues.merge" ( dict "values" ( list .Values.podLabels .Values.commonLabels ) "context" . ) }}
  selector: {{- include "common.labels.matchLabels" ( dict "customLabels" $podLabels "context" $ ) | nindent 4 }}
```

**Comment by cbrwflo** [comment]
Commit 25bb38d fixes the issue VIB caught (port items rendered without the `-`)

Without:
```yaml
# helm template --values .vib/logstash/runtime-parameters.yaml logstash ./bitnami/logstash > app.yaml
spec:
  type: ClusterIP
  clusterIP: None
  ports:
    name: http
    port: 8081
    protocol: TCP
    targetPort: http
    name: monitoring
    port: 80
    protocol: TCP
    targetPort: monitoring
```

With:
```yaml
spec:
  type: ClusterIP
  clusterIP: None
  ports:
    -
      name: http
      port: 8081
      protocol: TCP
      targetPort: http
    -
      name: monitoring
      port: 80
      protocol: TCP
      targetPort: monitoring
```

**Comment by cbrwflo** [comment]
I guess there's also `- {{ ... | toYaml | indent 6 | trim }}`, which looks like:

```yaml
spec:
  type: ClusterIP
  clusterIP: None
  ports:
    - name: http
      port: 8081
      protocol: TCP
      targetPort: http
    - name: monitoring
      port: 80
      protocol: TCP
      targetPort: monitoring
```

**Review by juan131 — APPROVED** [review]
LGTM

**Pull request merged**